kuebiko/pipeline/companies_house/flow.py

```diff
@@ -0,0 +1,188 @@
+        type=int,
+        default=6,
+    )
+
+    @step
+    def start(self):
+        """Parse parameters into URL's containing data chunks."""
+        ## It is important to **always have a docstring for `start` and `end`**
+        ## because their name cannot convey any information about what
+        ## the step does when looking at the DAG structure (without the code),
+        ## e.g. when using the `show` or `output-dot` commands.
+        import logging
+
+        self.urls = [
+            DATA_DUMP_URL_FORMAT.format(self.year, self.month, i, self.n_chunks)
+            for i in range(1, self.n_chunks + 1)
+        ]
+
+        if self.test_mode and not current.is_production:
+            self.urls = self.urls[:1]
+            logging.warning("TEST MODE: Constraining to first part of data!")
+
+        ## This step forks for each URL
+        self.next(self.get_data_chunk, foreach="urls")
+
+    @step
+    def get_data_chunk(self):
+        ## Due to the atomic nature of the following few steps in this flow,
+        ## docstrings would just repeat either/both of the step name or the
+        ## docstring of the single functions they tend to call.
+        from utils import read_from_url
+
+        self.raw = read_from_url(self.input)
+
+        ## Each fork (corresponding to one URL chunk), forks a further three times
+        self.next(self.process_organisation, self.process_address, self.process_sectors)
+
+    @step
+    def process_organisation(self):
+        from utils import process_organisations
+
+        self.organisations = process_organisations(self.raw)
+        self.next(self.join_branch)
+
+    @step
+    def process_address(self):
+        from utils import process_address
+
+        self.addresses = process_address(self.raw)
+        self.next(self.join_branch)
+
+    @step
+    def process_sectors(self):
+        from utils import process_sectors
+
+        self.sectors = process_sectors(self.raw)
+        self.next(self.join_branch)
+
+    ## Forks must be joined in reverse order
+    @step
+    def join_branch(self, inputs):
+        """Merge artifacts for a single data chunk."""
+        self.merge_artifacts(inputs)
+        self.next(self.join_foreach)
+
+    ## This step involves loading several large data artifacts; however there
+    ## is no way to separate this out into multiple extra steps to minimise RAM
+    ## because a join step must resolve all artifacts. The only solution is to
+    ## have a big enough machine.
+    @step
+    def join_foreach(self, inputs):
+        """Merge artifacts of different data chunks together."""
+        import logging
+
+        from pandas import concat
+
+        self.organisations = concat(input.organisations for input in inputs)
+        self.addresses = concat(input.addresses for input in inputs)
+        self.sectors = concat(input.sectors for input in inputs)
+
+        self.merge_artifacts(inputs, exclude=["raw"])
```

Comment:
@bishax: Just to make sure I understand: Is this line for removing `raw` variable from `self` and hence not persisting it in further steps and reducing the overhead?